Cubical/Experiments/List.agda

```diff
@@ -0,0 +1,204 @@
+
+ListEquiv : List A ≃ List' A
+ListEquiv = isoToEquiv ListIso
+
+-- We then use univalence to turn this into a path
+ListPath : (A : Type) → List A ≡ List' A
+ListPath A = isoToPath (ListIso {A = A})
+
+
+-- We can now use this path to transport the operations and properties
+-- from List to List'
+module transport where
+
+  -- First make a suitable Σ-type packaging what we need for the
+  -- transport (note that _++_ and rev here are part of the Σ-type).
+  -- It should be possible to automatically generate this given a module/file.
+  T : Type → Type
+  T X = Σ[ _++_ ∈ (X → X → X) ] Σ[ rev ∈ (X → X) ] ((xs ys : X) → rev (xs ++ ys) ≡ rev ys ++ rev xs)
+
+  -- We can now transport the instance of T for List to List'
+  T-List' : T (List' A)
+  T-List' {A = A} = transport (λ i → T (ListPath A i)) (_++_ , rev , rev-++-distr)
+
+  -- Getting the operations and property for List' is then just a matter of projecting them out
+  _++'_ : List' A → List' A → List' A
+  _++'_ = T-List' .fst
+
+  rev' : List' A → List' A
+  rev' = T-List' .snd .fst
+
+  rev-++-distr' : (xs ys : List' A) → rev' (xs ++' ys) ≡ rev' ys ++' rev' xs
+  rev-++-distr' = T-List' .snd .snd
+
+
+-- To connect this with the Cubical Agda paper consider the following
+-- (painfully) manual transport. This is really what the above code
+-- unfolds to.
+module manualtransport where
+
+  _++'_ : List' A → List' A → List' A
+  _++'_ {A = A} = transport (λ i → ListPath A i → ListPath A i → ListPath A i) _++_
+
+  rev' : List' A → List' A
+  rev' {A = A} = transport (λ i → ListPath A i → ListPath A i) rev
+
+  rev-++-distr' : (xs ys : List' A) → rev' (xs ++' ys) ≡ rev' ys ++' rev' xs
+  rev-++-distr' {A = A} = transport (λ i → (xs ys : ListPath A i)
+                                         → revP i (appP i xs ys) ≡ appP i (revP i ys) (revP i xs))
+                                    rev-++-distr
+    where
+    appP : PathP (λ i → ListPath A i → ListPath A i → ListPath A i) _++_ _++'_
+    appP i = transp (λ j → ListPath A (i ∧ j) → ListPath A (i ∧ j) → ListPath A (i ∧ j)) (~ i) _++_
+
+    revP : PathP (λ i → ListPath A i → ListPath A i) rev rev'
+    revP i = transp (λ j → ListPath A (i ∧ j) → ListPath A (i ∧ j)) (~ i) rev
+
+
+
+-- The above operations for List' are derived by going back and
+-- forth. With the SIP we can do better and transport properties for
+-- user defined operations (assuming that the operations are
+-- well-defined wrt to the forward direction of the equivalence).
+open import Cubical.Foundations.SIP
+open import Cubical.Structures.Axioms
+open import Cubical.Structures.Auto
+
+module SIP (A : Type) where
+
+  -- First define the raw structure without axioms. This is just the
+  -- signature of _++_ and rev.
+  RawStruct : Type → Type
+  RawStruct X = (X → X → X) × (X → X)
+
+  -- Some automated SIP magic
+  RawEquivStr : _
+  RawEquivStr = AutoEquivStr RawStruct
+
+  rawUnivalentStr : UnivalentStr _ RawEquivStr
+  rawUnivalentStr = autoUnivalentStr RawStruct
```

Comment:
@ecavallo Why is the "a" capitalized in `AutoEquivStr` but not in `autoUnivalentStr`?

Also, maybe these should be combined so that both get generated by one call to automation? I guess `transferAxioms` could take such a pair then

Reply:
I think the capitalized ones produce types and the lowercase ones don't?

Reply:
Carlo's right.

I'm not sure what will be the best way of combining record automation with the raw/induced structure stuff.